File: CodeArchieve/UOJ/Test-Round-1/83-水题出题人/problem_83/merge_sort.pas
program merge_sort;

var n, counter :longint;
var a :array[1..1000000] of longint;
var b, c :array[0..1000000] of longint;
var bn, cn :longint;

procedure merge_sort_r(ll, rr :longint);
var lr, rl, i :longint;
begin
	inc(counter);
	if ll < rr then begin
		inc(counter);
		lr := (ll + rr) div 2;
		inc(counter);
		rl := lr + 1;
		inc(counter);
		merge_sort_r(ll, lr);
		inc(counter);
		merge_sort_r(rl, rr);
		inc(counter);
		for i := ll to lr do begin
			inc(counter);
			b[i - ll] := a[i];
		end;
		inc(counter);
		b[lr - ll + 1] := maxlongint;
		inc(counter);
		for i := rl to rr do begin
			inc(counter);
			c[i - rl] := a[i];
		end;
		inc(counter);
		c[rr - rl + 1] := maxlongint;
		inc(counter);
		bn := 0;
		inc(counter);
		cn := 0;
		inc(counter);
		for i := ll to rr do begin
			inc(counter);
			if b[bn] < c[cn] then begin
				inc(counter);
				a[i] := b[bn];
				inc(counter);
				inc(bn);
			end else begin
				inc(counter);
				a[i] := c[cn];
				inc(counter);
				inc(cn);
			end;
		end;
	end;
end;

procedure merge_sort_main;
var i :longint;
begin
	inc(counter);
	readln(n);
	inc(counter);
	for i := 1 to n do begin
		inc(counter);
		readln(a[i]);
	end;
	inc(counter);
	merge_sort_r(1, n);
	inc(counter);
	for i := 1 to n do begin
		inc(counter);
		write(a[i], ' ');
	end;
end;

begin
	counter := 0;
	merge_sort_main;
	writeln;
	if counter <= 2000000 then begin
		writeln('Accepted! The counter value is :', counter);
	end else begin
		writeln('Time Limit Exceeded because the counter value is :', counter);
	end;
end.

File: CodeArchieve/UOJ/Test-Round-1/83-水题出题人/merge_sort.cpp
#include <stdio.h>

int n, counter;
int a[1000001];
int b[1000001], c[1000001];
int bn, cn;

void merge_sort_r(int ll, int rr){
	int lr, rl, i;
	counter++;
	if(ll < rr){
		counter++;
		lr = (ll + rr) / 2;
		counter++;
		rl = lr + 1;
		counter++;
		merge_sort_r(ll, lr);
		counter++;
		merge_sort_r(rl, rr);
		counter++;
		for(i = ll; i <= lr; i++){
			counter++;
			b[i - ll] = a[i];
		}
		counter++;
		b[lr - ll + 1] = 2147483647;
		counter++;
		for(i = rl; i <= rr; i++){
			counter++;
			c[i - rl] = a[i];
		}
		counter++;
		c[rr - rl + 1] = 2147483647;
		counter++;
		bn = 0;
		counter++;
		cn = 0;
		counter++;
		for(i = ll; i <= rr; i++){
			counter++;
			if(b[bn] < c[cn]){
				counter++;
				a[i] = b[bn];
				counter++;
				bn++;
			}else{
				counter++;
				a[i] = c[cn];
				counter++;
				cn++;
			}
		}
	}
}

void merge_sort_main(){
	int i;
	counter++;
	scanf("%d", &n);
	counter++;
	for(i = 1; i <= n; i++){
		counter++;
		scanf("%d", &a[i]);
	}
	counter++;
	merge_sort_r(1, n);
	counter++;
	for(i = 1; i <= n; i++){
		counter++;
		printf("%d ", a[i]);
	}
}

int main(){
	counter = 0;
	merge_sort_main();
	printf("\n");
	if(counter <= 2000000){
		printf("Accepted! The counter value is :%d\n", counter);
	}else{
		printf("Time Limit Exceeded because the counter value is :%d\n", counter);
	}
	return 0;
}
